### ISSUE TITLE
correct answer returned but incorrect generated code which override input df with random data

### ISSUE THREAD

**Issue by sandiemann** [gventuri/pandas-ai#139]
### 🐛 Describe the bug

I ran this on provided example,
```
dataframe = {
    "country": [
        "United States",
        "United Kingdom",
        "France",
        "Germany",
        "Italy",
        "Spain",
        "Canada",
        "Australia",
        "Japan",
        "China",
    ],
    "gdp": [
        19294482071552,
        2891615567872,
        2411255037952,
        3435817336832,
        1745433788416,
        1181205135360,
        1607402389504,
        1490967855104,
        4380756541440,
        14631844184064,
    ],
    "happiness_index": [6.94, 7.16, 6.66, 7.07, 6.38, 6.4, 7.23, 7.22, 5.87, 5.12],
}

import pandas as pd
  ...: 
  ...: from pandasai import PandasAI
  ...: from pandasai.llm.openai import OpenAI
  ...: 
  ...: from examples.data.sample_dataframe import dataframe
  ...: 
  ...: df = pd.DataFrame(dataframe)
  ...: 
  ...: llm = OpenAI()
  ...: pandas_ai = PandasAI(llm, verbose=True, conversational=False)
  ...: response = pandas_ai.run(df, "Calculate the sum of the gdp of north american countries")
  ...: print(response)
  ...: 
Running PandasAI with openai LLM...
testing...
Code generated:

import pandas as pd
# create dataframe
data = {'country': ['France', 'Germany', 'United States', 'Italy', 'United Kingdom', 'Canada', 'Mexico', 'Brazil', 'Argentina', 'Chile'],
        'gdp': [2411255037952, 1726692115, 2269578488, 8119293427, 1371148559, 1670186276, 1218570714, 2143320000, 518027000, 247680000],
        'happiness_index': [6.38, 7.07, 7.16, 6.38, 7.16, 7.23, 6.47, 6.95, 6.39, 6.44]}
df = pd.DataFrame(data)
# calculate sum of gdp for North American countries
north_america = ['United States', 'Canada', 'Mexico']
sum_gdp = df[df['country'].isin(north_america)]['gdp'].sum()
print(sum_gdp)

Answer: 20901884461056
20901884461056

```

As, the returned answer (USA + Canada GDP sum) is correct but the code is completely wrong as it discarded the input dataframe and generated random dataframe.

**Comment by QJ-Chen** [comment]
The "code generated" in the log seems to be the code generated for the first time. There were errors when executing as it's wrong so that the error_correction_framework will be applied for `_max_retries` times. See `pandas_ai.run_code` in the llm/\_\_init\_\_.py (Though there seems to be some problems with the error_correction I think.)

**Comment by gventuri** [comment]
@sandiemann thanks for reporting. This is in fact a well-known hallucination that we fixed by manually removing every override of `df`. We print the generated code, not the one that we run, and this is in fact very misleading.

Should be fixed with the new commit [cbd9aa8](https://github.com/gventuri/pandas-ai/commit/cbd9aa89d8551780c48aacdea8094754f7c13304)

**Comment by sandiemann** [comment]
@gventuri I think the issue is when returning the code, the llm generated code creates new dataframe and exclude the one which was provided.  The way to explicitly run operation on the provided df, we need to pass it in the prompt to generate code only based on this dataframe and do not create a new dataframe.  I was trying to fix here in [PR](https://github.com/gventuri/pandas-ai/pull/140) commit (https://github.com/gventuri/pandas-ai/pull/140/commits/780b8af74ee009360b69cda67dd315e60b2b2163)
